File: tree/111.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type Node struct {
    level int
    Node *TreeNode
}

// bfs
func minDepth(root *TreeNode) int {
    if root == nil {
        return 0
    }
    l := make([]*Node, 0)
    l = append(l, &Node{0, root})
    var t *Node
    for len(l) > 0 {
        t = l[0]
        if t.Node.Left == nil && t.Node.Right == nil {
            return t.level + 1
        }
        l = l[1:]
        if t.Node.Left != nil {
            l = append(l, &Node{t.level+1, t.Node.Left})
        }
        if t.Node.Right != nil {
            l = append(l, &Node{t.level+1, t.Node.Right})
        }
    }
    return 0   
}

File: tree/111.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (!root)
            return 0;
        int ret = INT_MAX;
        dfs(root, ret, 0);
        return ret;
    }
    void dfs(TreeNode* root, int& ret, int cur) {
        if (!root->left && !root->right) {
            if ((cur+1) < ret) {
                ret = cur + 1;
            }
        }
        if (root->left)
            dfs(root->left, ret, cur+1);
        if (root->right)
            dfs(root->right, ret, cur+1);
    }
};

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if (!root)
            return 0;
        int ret = INT_MAX;
        dfs(root, ret, 0);
        return ret;
    }
    void dfs(TreeNode* root, int& ret, int cur) {
        if (!root->left && !root->right) {
            if ((cur+1) < ret) {
                ret = cur + 1;
            }
        }
        if (root->left)
            dfs(root->left, ret, cur+1);
        if (root->right)
            dfs(root->right, ret, cur+1);
    }
};
